_Lambda School Data Science, Unit 2_
 
# Classification 1 Sprint Challenge: Predict Steph Curry's shots 🏀

For your Sprint Challenge, you'll use a dataset with all Steph Curry's NBA field goal attempts. (Regular season and playoff games, from October 28, 2009, through June 5, 2019.) 

You'll use information about the shot and the game to predict whether the shot was made. This is hard to predict! Try for an accuracy score in the high 50's or low 60's. The dataset was collected with the [nba_api](https://github.com/swar/nba_api) Python library.

In [29]:
import pandas as pd
url = 'https://drive.google.com/uc?export=download&id=1fL7KPyxgGYfQDsuJoBWHIWwCAf-HTFpX'
df = pd.read_csv(url, parse_dates=['game_date']).set_index('game_date')
assert df.shape == (13958, 19)

In [30]:
# Looking at the data

df.head()

game_id  game_event_id    player_name  period  minutes_remaining  \
game_date                                                                       
2009-10-28  20900015              4  Stephen Curry       1                 11   
2009-10-28  20900015             17  Stephen Curry       1                  9   
2009-10-28  20900015             53  Stephen Curry       1                  6   
2009-10-28  20900015            141  Stephen Curry       2                  9   
2009-10-28  20900015            249  Stephen Curry       2                  2   

            seconds_remaining          action_type       shot_type  \
game_date                                                            
2009-10-28                 25            Jump Shot  3PT Field Goal   
2009-10-28                 31  Step Back Jump shot  2PT Field Goal   
2009-10-28                  2            Jump Shot  2PT Field Goal   
2009-10-28                 49            Jump Shot  2PT Field Goal   
2009-10-28                 19            Jump Shot  2PT Field Goal   

                  shot_zone_basic         shot_zone_area shot_zone_range  \
game_date                                                                  
2009-10-28      Above the Break 3  Right Side Center(RC)         24+ ft.   
2009-10-28              Mid-Range   Left Side Center(LC)       16-24 ft.   
2009-10-28  In The Paint (Non-RA)              Center(C)        8-16 ft.   
2009-10-28              Mid-Range           Left Side(L)       16-24 ft.   
2009-10-28              Mid-Range   Left Side Center(LC)       16-24 ft.   

            shot_distance  loc_x  loc_y  shot_made_flag  htm  vtm  \
game_date                                                           
2009-10-28             26     99    249               0  GSW  HOU   
2009-10-28             18   -122    145               1  GSW  HOU   
2009-10-28             14    -60    129               0  GSW  HOU   
2009-10-28             19   -172     82               0  GSW  HOU   
2009-10-28             16    -68    148               0  GSW  HOU   

               season_type  scoremargin_before_shot  
game_date                                            
2009-10-28  Regular Season                      2.0  
2009-10-28  Regular Season                      0.0  
2009-10-28  Regular Season                     -4.0  
2009-10-28  Regular Season                     -4.0  
2009-10-28  Regular Season                      0.0

In [18]:
# Putting years into a feature
# Didnt work for some reason

# train = df[df['game_date'] < 2017]

In [11]:
df.head()

game_id  game_event_id    player_name  period  minutes_remaining  \
game_date                                                                       
2009-10-28  20900015              4  Stephen Curry       1                 11   
2009-10-28  20900015             17  Stephen Curry       1                  9   
2009-10-28  20900015             53  Stephen Curry       1                  6   
2009-10-28  20900015            141  Stephen Curry       2                  9   
2009-10-28  20900015            249  Stephen Curry       2                  2   

            seconds_remaining          action_type       shot_type  \
game_date                                                            
2009-10-28                 25            Jump Shot  3PT Field Goal   
2009-10-28                 31  Step Back Jump shot  2PT Field Goal   
2009-10-28                  2            Jump Shot  2PT Field Goal   
2009-10-28                 49            Jump Shot  2PT Field Goal   
2009-10-28                 19            Jump Shot  2PT Field Goal   

                  shot_zone_basic         shot_zone_area shot_zone_range  \
game_date                                                                  
2009-10-28      Above the Break 3  Right Side Center(RC)         24+ ft.   
2009-10-28              Mid-Range   Left Side Center(LC)       16-24 ft.   
2009-10-28  In The Paint (Non-RA)              Center(C)        8-16 ft.   
2009-10-28              Mid-Range           Left Side(L)       16-24 ft.   
2009-10-28              Mid-Range   Left Side Center(LC)       16-24 ft.   

            shot_distance  loc_x  loc_y  shot_made_flag  htm  vtm  \
game_date                                                           
2009-10-28             26     99    249               0  GSW  HOU   
2009-10-28             18   -122    145               1  GSW  HOU   
2009-10-28             14    -60    129               0  GSW  HOU   
2009-10-28             19   -172     82               0  GSW  HOU   
2009-10-28             16    -68    148               0  GSW  HOU   

               season_type  scoremargin_before_shot  year  
game_date                                                  
2009-10-28  Regular Season                      2.0  1970  
2009-10-28  Regular Season                      0.0  1970  
2009-10-28  Regular Season                     -4.0  1970  
2009-10-28  Regular Season                     -4.0  1970  
2009-10-28  Regular Season                      0.0  1970

In [25]:
from sklearn.model_selection import train_test_split

In [28]:
# I don't know why this didn't work

# df['date'] = pd.to_datetime(df['game_date']).dt.year

In [31]:
import datetime as dt

In [34]:
# data_df['Date'] = pd.to_datetime(df['Date'])
# data_df['Date']=data_df['Date'].map(dt.datetime.toordinal)

In [20]:
import category_encoders as ce
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [ ]:
train = df[df['game_date']]

In [33]:
# train, val, test = train_test_split(df, train_size=0.80, test_size=0.20,
#                              stratify=df['shot_made_flag'])


In [36]:
# train, val = train_test_split(train, train_size=0.80, test_size=0.20, 
#                               stratify=train['status_group'], random_state=42)

In [38]:
train = df.loc['2009-10' : '2017-6']
train.shape
val = df.loc['2017-10' : '2018-6']
val.shape
test = df.loc['2018-10' : '2019-6']
test.shape

(1709, 19)

In [40]:
X_train = train
y_train = train['shot_made_flag']
X_train, X_val, y_train, y_val = train_test_split(X_train, y_train,
                                 train_size=0.80, test_size=0.20,
                                 stratify=y_train)

X_train.shape, X_val.shape, y_train.shape, y_val.shape

((8864, 19), (2217, 19), (8864,), (2217,))

In [42]:
features = ['minutes_remaining', 'seconds_remaining', 
            'action_type', 'shot_type']

X_train_subset = X_train[features]
X_val_subset = X_val[features]

encoder = ce.OneHotEncoder(use_cat_names=True)
X_train_encoded = encoder.fit_transform(X_train_subset)
X_val_encoded = encoder.transform(X_val_subset)

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train_encoded)
X_val_scaled = scaler.transform(X_val_encoded)

model = LogisticRegression(solver='lbfgs', multi_class='auto', max_iter=1000)
model.fit(X_train_scaled, y_train)
print('validation accuracy', model.score(X_val_scaled, y_val))

validation accuracy 0.6806495263870095


/Users/mattmastin/anaconda3/lib/python3.7/site-packages/sklearn/preprocessing/data.py:645: DataConversionWarning: Data with input dtype int64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
/Users/mattmastin/anaconda3/lib/python3.7/site-packages/sklearn/base.py:464: DataConversionWarning: Data with input dtype int64 were all converted to float64 by StandardScaler.
  return self.fit(X, **fit_params).transform(X)
/Users/mattmastin/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:13: DataConversionWarning: Data with input dtype int64 were all converted to float64 by StandardScaler.
  del sys.path[0]


This Sprint Challenge has two parts. To demonstrate mastery on each part, do all the required, numbered instructions. To earn a score of "3" for the part, also do the stretch goals.

## Part 1. Prepare to model

### Required

1. **Do train/validate/test split.** Use the 2009-10 season through 2016-17 season to train, the 2017-18 season to validate, and the 2018-19 season to test. NBA seasons begin in October and end in June. You'll know you've split the data correctly when your train set has 11081 observations, your validation set has 1168 observations, and your test set has 1709 observations.
2. **Begin with baselines for classification.** Your target to predict is `shot_made_flag`. What is the baseline accuracy for the validation set, if you guessed the majority class for every prediction?
3. **Use Ordinal Encoding _or_ One-Hot Encoding,** for the categorical features you select.
4. **Train a Random Forest _or_ Logistic Regression** with the features you select.

### Stretch goals
Engineer at least 4 of these 5 features:

- **Homecourt Advantage**: Is the home team (`htm`) the Golden State Warriors (`GSW`) ?
- **Opponent**: Who is the other team playing the Golden State Warriors?
- **Seconds remaining in the period**: Combine minutes remaining with seconds remaining, to get the total number of seconds remaining in the period.
- **Seconds remaining in the game**: Combine period, and seconds remaining in the period, to get the total number of seconds remaining in the game. A basketball game has 4 periods, each 12 minutes long.
- **Made previous shot**: Was Steph Curry's previous shot successful?
    

## Part 2. Evaluate models

### Required
1. Get your model's **validation accuracy.** (Multiple times if you try multiple iterations.)
2. Get your model's **test accuracy.** (One time, at the end.)
3. Get and plot your Random Forest's **feature importances** _or_ your Logistic Regression's **coefficients.**
4. Imagine this is the confusion matrix for a binary classification model. **Calculate accuracy, precision, and recall for this confusion matrix:**

<table>
  <tr>
    <td colspan="2" rowspan="2"></td>
    <td colspan="2">Predicted</td>
  </tr>
  <tr>
    <td>Negative</td>
    <td>Positive</td>
  </tr>
  <tr>
    <td rowspan="2">Actual</td>
    <td>Negative</td>
    <td style="border: solid">85</td>
    <td style="border: solid">58</td>
  </tr>
  <tr>
    <td>Positive</td>
    <td style="border: solid">8</td>
    <td style="border: solid"> 36</td>
  </tr>
</table>


### Stretch goals
- Calculate F1 score for the provided, imaginary confusion matrix.
- Plot a real confusion matrix for your basketball model, with row and column labels.
- Print the classification report for your model.